tensorboard/plugins/hparams/api.py

```diff
@@ -0,0 +1,378 @@
+    if not isinstance(max_value, float):
+      raise TypeError("max_value must be a float: %r" % (max_value,))
+    if min_value > max_value:
+      raise ValueError("%r > %r" % (min_value, max_value))
+    self._min_value = min_value
+    self._max_value = max_value
+
+  def __str__(self):
+    return "[%s, %s]" % (self._min_value, self._max_value)
+
+  def __repr__(self):
+    return "RealInterval(%r, %r)" % (self._min_value, self._max_value)
+
+  @property
+  def dtype(self):
+    return float
+
+  @property
+  def min_value(self):
+    return self._min_value
+
+  @property
+  def max_value(self):
+    return self._max_value
+
+  def update_hparam_info(self, hparam_info):
+    hparam_info.type = api_pb2.DATA_TYPE_FLOAT64
+    hparam_info.domain_interval.min_value = self._min_value
+    hparam_info.domain_interval.max_value = self._max_value
+
+
+class Discrete(Domain):
+  """A domain that takes on a fixed set of values.
+
+  These values may be of any (single) domain type.
+  """
+
+  def __init__(self, values, dtype=None):
+    self._values = list(values)
+    if dtype is None:
+      if self._values:
+        dtype = type(self._values[0])
+      else:
+        raise ValueError("Empty domain with no dtype specified")
+    if dtype not in (int, float, bool, str):
+      raise ValueError("Unknown dtype: %r" % (dtype,))
+    self._dtype = dtype
+    for value in self._values:
+      if not isinstance(value, self._dtype):
+        raise ValueError(
+            "dtype mismatch: not isinstance(%r, %s)"
+            % (value, self._dtype.__name__)
+        )
+    self._values.sort()
+
+  def __str__(self):
+    return "{%s}" % (", ".join(repr(x) for x in self._values))
+
+  def __repr__(self):
+    return "Discrete(%r)" % (self._values,)
+
+  @property
+  def dtype(self):
+    return self._dtype
+
+  @property
+  def values(self):
+    return list(self._values)
+
+  def update_hparam_info(self, hparam_info):
+    hparam_info.type = {
+        int: api_pb2.DATA_TYPE_FLOAT64,  # TODO(#1998): Add int dtype.
+        float: api_pb2.DATA_TYPE_FLOAT64,
+        bool: api_pb2.DATA_TYPE_BOOL,
+        str: api_pb2.DATA_TYPE_STRING,
+    }[self._dtype]
+    hparam_info.ClearField("domain_discrete")
+    hparam_info.domain_discrete.extend(self._values)
+
+
+class Metric(object):
+  """A metric in an experiment.
+
+  A metric is a real-valued function of a model. Each metric is
+  associated with a TensorBoard scalar summary, which logs the metric's
+  value as the model trains.
+  """
+  TRAINING = api_pb2.DATASET_TRAINING
+  VALIDATION = api_pb2.DATASET_VALIDATION
+
+  def __init__(
+      self,
+      tag,
+      group=None,
+      display_name=None,
+      description=None,
+      dataset_type=None,
+  ):
+    """
+    Args:
+      tag: The tag name of the scalar summary that corresponds to this
+        metric (as a `str`).
+      group: An optional string listing the subdirectory under the
```

Comment:
Usage question: so I want metric to be written in arbitrary deep path, I should pass a string with the correct path separator for a specific OS?